internal/descriptor/provider/provider.go

```diff
@@ -1,109 +1,90 @@
+// that already have all required data.
+// Then we don't have to create intermediate variables of type ocmidentity.Component.
+type OCMIProvider interface {
+	GetOCMIdentity() (*ocmidentity.Component, error)
+}
 
-		if desc == nil {
-			return nil, ErrDescriptorNil
-		}
+func (c *CachedDescriptorProvider) Add(ocmi ocmidentity.Component) error {
+	_, err := c.getDescriptor(ocmi, true)
+	return err
+}
 
-		return desc, nil
-	}
-	key := c.GenerateDescriptorKey(template.Name, template.Spec.Version)
+func (c *CachedDescriptorProvider) GetDescriptor(ocmi ocmidentity.Component) (*types.Descriptor, error) {
+	return c.getDescriptor(ocmi, false)
+}
 
-	descriptor := c.DescriptorCache.Get(key)
-	if descriptor != nil {
-		return descriptor, nil
+func (c *CachedDescriptorProvider) GetDescriptorWithIdentity(ocp OCMIProvider) (*types.Descriptor, error) {
+	if ocp == nil {
+		return nil, fmt.Errorf("failed to get component identity from provider: %w", ErrNilProvider)
 	}
 
-	ocmDesc, err := compdesc.Decode(
-		template.Spec.Descriptor.Raw, []compdesc.DecodeOption{compdesc.DisableValidation(true)}...,
-	)
+	ocmi, err := ocp.GetOCMIdentity()
 	if err != nil {
-		return nil, errors.Join(ErrDecode, err)
+		return nil, fmt.Errorf("failed to get component identity from provider: %w", err)
 	}
-
-	template.Spec.Descriptor.Object = &types.Descriptor{ComponentDescriptor: ocmDesc}
-	descriptor, ok := template.Spec.Descriptor.Object.(*types.Descriptor)
-	if !ok {
-		return nil, ErrTypeAssert
+	if ocmi == nil {
+		return nil, fmt.Errorf("failed to get component identity from provider: %w", ErrNilIdentity)
 	}
 
-	return descriptor, nil
-}
-
-func (c *CachedDescriptorProvider) GenerateDescriptorKey(name, version string) string {
-	return fmt.Sprintf("%s:%s", name, version)
+	return c.getDescriptor(*ocmi, false)
 }
 
-func (c *CachedDescriptorProvider) Add(template *v1beta2.ModuleTemplate) error {
-	if template == nil {
-		return ErrTemplateNil
+func (c *CachedDescriptorProvider) getDescriptor(ocmi ocmidentity.Component, updateCache bool) (
+	*types.Descriptor, error,
+) {
+	if ocmi.Name() == "" || ocmi.Version() == "" {
+		return nil, fmt.Errorf("cannot get descriptor for component: %w", ErrNameOrVersionEmpty)
 	}
-
-	key := c.GenerateDescriptorKey(template.Name, template.Spec.Version)
-	descriptor := c.DescriptorCache.Get(key)
+	key := cache.GenerateDescriptorKey(ocmi)
+	descriptor := c.descriptorCache.Get(key)
 	if descriptor != nil {
-		return nil
-	}
-
-	if template.Spec.Descriptor.Object != nil {
-		desc, ok := template.Spec.Descriptor.Object.(*types.Descriptor)
-		if ok && desc != nil {
-			c.DescriptorCache.Set(key, desc)
-			return nil
-		}
+		return descriptor, nil
 	}
 
-	ocmDesc, err := compdesc.Decode(
-		template.Spec.Descriptor.Raw, []compdesc.DecodeOption{compdesc.DisableValidation(true)}...,
-	)
+	descriptor, err := c.descriptorService.GetComponentDescriptor(context.Background(), ocmi)
```

Comment:
You should not use background context here without cancelling.

// Background returns a non-nil, empty [Context]. It is never canceled, has no
// values, and has no deadline. It is typically used by the main function,
// initialization, and tests, and as the top-level Context for incoming
// requests.

the context here should have `WithCancel`, and get cancelled after function call.

```
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
```